File: src/Agenda.Frontend/src/models/page-of.ts
/**
 * Represents a paginated response from the API.
 */
export interface PageOf<T> {
  /** Current page number (1-indexed) */
  page: number;

  /** Total number of pages */
  total: number;

  /** Number of items in the current page */
  count: number;

  /** Requested page size (when provided by backend) */
  pageSize?: number;

  /** Total number of matching items across all pages (when provided by backend) */
  totalCount?: number;

  /** Items in the current page */
  items: T[];

  /** Navigation links */
  links: PageLinks;
}

/**
 * Represents pagination links for navigation
 */
export interface PageLinks {
  /** Link to the first page */
  first?: PageLink;

  /** Link to the last page */
  last?: PageLink;

  /** Link to the previous page */
  previous?: PageLink;

  /** Link to the next page */
  next?: PageLink;
}

/**
 * Represents a single pagination link
 */
export interface PageLink {
  href: string;
  relations: string[];
}

File: src/Agenda.API/Features/PageOf.cs
using System.Collections.Generic;
using System.Linq;

namespace Agenda.API.Features;

/// <summary>
/// Wraps a page
/// </summary>
/// <typeparam name="TResource">Type of the resource the page will contain</typeparam>
public class PageOf<TResource> where TResource : class
{
    /// <summary>
    /// Index of the page
    /// </summary>
    public int Page { get; init; }

    /// <summary>
    /// Indicates the number of <typeparamref name="TResource"/> elements the current page is a subset of
    /// </summary>
    public long Total { get; init; }

    /// <summary>
    /// Size requested for each page.
    /// </summary>
    public int PageSize { get; init; }

    /// <summary>
    /// Total number of <typeparamref name="TResource"/> elements matching the criteria.
    /// </summary>
    public long TotalCount { get; init; }

    /// <summary>
    /// The number of <typeparamref name="TResource"/> elements the current page holds
    /// </summary>
    public long Count { get; init; }

    /// <summary>
    /// <typeparamref name="TResource"/> elements held in the current page
    /// </summary>
    public IEnumerable<TResource> Items
    {
        get => _items;
        set => _items = value ?? []; // Avoid null reference exceptions when enumerating over items
    }

    private IEnumerable<TResource> _items;

    /// <summary>
    /// Navigation links between pages result
    /// </summary>
    public PageLinks Links { get; init; }

    /// <summary>
    /// Builds an empty page
    /// </summary>
    public PageOf()
    {
        _items = new List<TResource>();
    }
}